### PR TITLE
Signed comparator gadget: SLT, SGT

### PR DESCRIPTION
**Pull request by scroll-dev**
Implement the circuit for SLT / SGT. The spec PR is at https://github.com/appliedzkp/zkevm-specs/pull/95

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _transaction: &Transaction<F>,
+        _call: &Call<F>,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let opcode = step.opcode.unwrap();
```
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs:161]: Why is `opcode` an `Option` here? 
If you have a step there's no way on which it does not contain an opcode no?

File: zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+            region,
+            offset,
+            F::from(a[31] as u64),
+            F::from(128u64),
+        )?;
+        self.sign_check_b.assign(
+            region,
+            offset,
+            F::from(b[31] as u64),
+            F::from(128u64),
```
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs:190]: Sorry to ask. But unsure why `rhs` is passed as 128.

**Review by CPerezz — COMMENTED** [review]
Amazing job. Brilliant PR with really really nice documentation!

Just two questions. But overall LGTM! :+1:

File: zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs (hunk 1 above)
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs:161]: We could have internal `ExecutionState` such as `BeginTx` that doesn't have any corresponding opcode. That's why it's an option field in the `ExecStep`.

Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs:161]: Ohhh that's right!! Thanks!!! :)

File: zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs (hunk 2 above)
Comment (roynalnaruto) [reply zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs:190]: @CPerezz The values `a` and `b` are in the two's complement form, representing signed integers. So for the most significant byte (MSB), i.e. `a[31]` or `b[31]`, the first bit is set to `1` to indicate that the value `a` or `b` is negative.

If that bit is `1`, it means the `MSB >= 128`. So by checking if `MSB < 128`, we know that the value of `a` or `b` itself is `>= 0`.

Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs:190]: That makes complete sense! 
Apologies for the question :sweat_smile:

**Review by CPerezz — APPROVED** [review]
LGTM!

@ed255 @han0110 can we get one more review from any of you so that we can merge this?

**Comment by ed255** [comment]
> LGTM!
> 
> @ed255 @han0110 can we get one more review from any of you so that we can merge this?

Let me take a look at https://github.com/appliedzkp/zkevm-specs/pull/95 first and then I'll review this :)

**Review by ed255 — APPROVED** [review]
LGTM!

**Comment by ed255** [comment]
Although this has 2 approvals, I'd prefer to wait until the spec https://github.com/appliedzkp/zkevm-specs/pull/95 is merged to merge this PR.

**Pull request merged**